### PR TITLE
Debtcache import excluded debt entries

### PR DESCRIPTION
**Pull request by artdgn**
Adds the ability to import excluded debt entries which is needed for upgrading to new DebtCache instance.

### CODE REVIEW

File: contracts/BaseDebtCache.sol [hunk 1]
```diff
@@ -31,6 +31,9 @@ contract BaseDebtCache is Owned, MixinSystemSettings, IDebtCache {
     uint internal _cacheTimestamp;
     bool internal _cacheInvalid = true;
 
+    // flag to ensure importing excluded debt is invoked only once
+    bool public excludedDebtImported = false; // public to avoid needing an event
```
Comment (dbeal-eth) [inline contracts/BaseDebtCache.sol:35]: np, i'd just call it `isInitialized` if you really want to be future proof lol

File: contracts/BaseDebtCache.sol [hunk 2]
```diff
@@ -204,6 +207,34 @@ contract BaseDebtCache is Owned, MixinSystemSettings, IDebtCache {
...
+    /// used when migrating to new DebtCache instance in order to import the excluded debt records
+    /// If this method is not run after upgrading the contract, the debt will be
+    /// incorrect w.r.t to wrapper factory assets until the values are imported from
+    /// previous instance of the contract
+    /// Also, in addition to this method it's possible to use recordExcludedDebtChange since
+    /// it's accessible to owner in case additional adjustments are required
+    function importExcludedIssuedDebts(IDebtCache prevDebtCache) external onlyOwner {
+        // this can only be run once so that recorded debt deltas aren't accidentally
+        // lost or double counted
+        require(excludedDebtImported == false, "import can only be run once");
```
Comment (dbeal-eth) [inline contracts/BaseDebtCache.sol:219]: np `excludedDebtImported == false` should be replaced with `!excludedDebtImported`

js linters would throw a hissie fit about this

**Review by dbeal-eth — APPROVED** [review]
nice

File: contracts/BaseDebtCache.sol (hunk 1 above)
Comment (jacko125) [reply contracts/BaseDebtCache.sol:35]: We do use this pattern elsewhere before of `isInitialized`

File: contracts/BaseDebtCache.sol [hunk 3]
```diff
@@ -204,6 +207,34 @@ contract BaseDebtCache is Owned, MixinSystemSettings, IDebtCache {
...
+        excludedDebtImported = true;
+
+        bytes32[] memory currencyKeys = issuer().availableCurrencyKeys();
```
Comment (jacko125) [inline contracts/BaseDebtCache.sol:222]: So here we require the synths to be in the issuer() first before executing this function?  Ie, if new issuer is deployed we need to add synths into it first before invoking this function.

**Review by jacko125 — APPROVED** [review]

File: contracts/BaseDebtCache.sol (hunk 1 above)
Comment (artdgn) [reply contracts/BaseDebtCache.sol:35]: fair point, but I wanted to be specific and save an event :)

File: contracts/BaseDebtCache.sol (hunk 3 above)
Comment (artdgn) [reply contracts/BaseDebtCache.sol:222]: good point, will add a check for this in the script to ensure new issuer has synths.

Comment (artdgn) [reply contracts/BaseDebtCache.sol:222]: actually, better to just import using the keys from the previous issuer

**Review by dbeal-eth — APPROVED** [review]

**Review by jacko125 — APPROVED** [review]
LGTM!

**Pull request merged**